library/core/src/alloc/global.rs

```diff
@@ -53,6 +53,29 @@ use crate::ptr;
 /// * `Layout` queries and calculations in general must be correct. Callers of
 ///   this trait are allowed to rely on the contracts defined on each method,
 ///   and implementors must ensure such contracts remain true.
+///
+/// * You may not rely on allocations actually happening, even if there are explicit
+///   heap allocations in the source.
+///   The optimizer may detect unused allocations that it can either
+///   eliminate entirely or move to the stack and thus never invoke the allocator here. The
+///   optimizer may further assume that allocation is infallible, so code that used to fail due
+///   to allocator failures may now suddenly work because the optimizer worked around the
+///   need for an allocation.
+///   More concretely, the following code example is unsound, irrespective of whether your
```

Comment:
Some of the wrapping in this PR feels a bit suspect; if you want lines like this to start new paragraphs, you need to have an empty line between them. This will coalesce into one big paragraph otherwise. If the intent is *not* for multiple paragraphs, then this should be reformatted too.

You can use https://spec.commonmark.org/dingus/ to easily check how things render, if you remove the `/// `s first, of course.

Reply:
:shrug: I have no strong opinion. I know how it renders, my markdown is usually just completely random in where I use non-formatting whitespace. I'll make sure it's closer to how it'll render